x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java

```diff
@@ -397,266 +289,102 @@ protected void doSaveState(IndexerState indexerState, TransformIndexerPosition p
+                                nil -> {
+                                    logger.trace("[{}] deleted old transform stats and state document", getJobId());
+                                    listener.onResponse(null);
+                                },
+                                e -> {
+                                    String msg = LoggerMessageFormat.format(
+                                        "[{}] failed deleting old transform configurations.",
+                                        getJobId()
+                                    );
+                                    logger.warn(msg, e);
+                                    // If we have failed, we should attempt the clean up again later
+                                    oldStatsCleanedUp.set(false);
+                                    listener.onResponse(null);
+                                }
+                            )
+                        );
+                    } else {
+                        listener.onResponse(null);
+                    }
                 },
-                e -> {
-                    logger.warn(
+                statsExc -> {
+                    logger.error(
                         new ParameterizedMessage(
-                            "[{}] failed to detect changes for transform. Skipping update till next check.",
-                            getJobId()),
-                        e);
-                    auditor.warning(getJobId(),
-                        "Failed to detect changes for transform, skipping update till next check. Exception: "
-                            + e.getMessage());
-                    hasChangedListener.onResponse(false);
-                }));
-    }
-
-    private boolean isIrrecoverableFailure(Exception e) {
-        return e instanceof IndexNotFoundException
-            || e instanceof AggregationResultUtils.AggregationExtractionException
-            || e instanceof TransformConfigReloadingException;
-    }
+                            "[{}] updating stats of transform failed.",
+                            transformConfig.getId()
+                        ),
+                        statsExc
+                    );
+                    auditor.warning(
+                        getJobId(),
+                        "Failure updating stats of transform: " + statsExc.getMessage()
+                    );
+                    // for auto stop shutdown the task
+                    if (state.getTaskState().equals(TransformTaskState.STOPPED)) {
+                        context.shutdown();
+                    }
+                    listener.onFailure(statsExc);
+                }
+            )
+        );
 
-    synchronized void handleFailure(Exception e) {
-        logger.warn(new ParameterizedMessage("[{}] transform encountered an exception: ",
-            getJobId()),
-            e);
-        if (handleCircuitBreakingException(e)) {
-            return;
+        if (shouldStopAtCheckpoint) {
```

Comment:
Having this at the end of `doSaveState` does not make sense. This means that if `shouldStopAtCheckpoint` is true, then we will stop at the end of a `doSaveState` call. But, `doSaveState` can be called once we process 50 pages and is no indication of actually finishing a checkpoint or not. 

In short, how do you know this is the `doSaveState` after a checkpoint has finished?

Reply:
I am not aware that I changed the logic, maybe this is a unclean merge? Will check.

Reply:
The do private doSaveState: https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java#L413

There are no checks for `shouldStopAtCheckpoint`. It might be a weird merge for sure.